### PR TITLE
auto-trigger a/b

### PR DESCRIPTION
**Pull request by andrewyuq**
1. The configurable actions are added to the official release regardless(i.e. seeing them in the keymap), the difference is that for treatment group users they will be able to use the new configurable actions while the control group won't be able to use them.
2. Create mutliple `xxxNew` classes that will be used for new auto-trigger UX, old UX will continue to use the classes without the `New` suffix. 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist
- [ ] My code follows the code style of this project
- [ ] I have added tests to cover my changes
- [ ] A short description of the change has been added to the **[CHANGELOG](https://github.com/aws/aws-toolkit-jetbrains/blob/master/CONTRIBUTING.md#contributing-via-pull-requests)** if the change is customer-facing in the IDE.
- [ ] I have added metrics for my changes (if required)
 
## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/editor/CodeWhispererEditorListener.kt [hunk 1]
```diff
@@ -29,7 +31,11 @@ class CodeWhispererEditorListener : EditorFactoryListener {
                 // the most accurate code percentage data.
                 override fun documentChanged(event: DocumentEvent) {
                     if (!isCodeWhispererEnabled(project)) return
-                    CodeWhispererInvocationStatus.getInstance().documentChanged()
+                    if (CodeWhispererFeatureConfigService.getInstance().getNewAutoTriggerUX()) {
+                        CodeWhispererInvocationStatusNew.getInstance().documentChanged()
+                    } else {
+                        CodeWhispererInvocationStatus.getInstance().documentChanged()
```
Comment (Will-ShaoHua) [inline plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/editor/CodeWhispererEditorListener.kt:37]: for all these singleton service, I think we can do 
```
fun getInstance(): CodeWhispererService = if (newUX) CodeWhispererServiceNew.getInstance() else service()
```

File: plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/editor/CodeWhispererEditorManagerNew.kt [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+        val oldOffset = caretPosition.offset
+        val newOffset = editor.caretModel.primaryCaret.offset
+        return when {
+            oldOffset < newOffset -> CaretMovement.MOVE_FORWARD
+            oldOffset > newOffset -> CaretMovement.MOVE_BACKWARD
+            else -> CaretMovement.NO_CHANGE
+        }
+    }
+
+    fun getMatchingSymbolsFromRecommendation(
```
Comment (rli) [inline plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/editor/CodeWhispererEditorManagerNew.kt:101]: if it's not too hard can you pull out the common code so there's less delta and risk

**Review by rli — APPROVED** [review]
i see lots of code that appears duplicate between old/new so it is difficult to figure out what is new and what is just the same logic but referencing new objects

**Review by Will-ShaoHua — APPROVED** [review]

File: plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererTelemetryServiceNew.kt [hunk 3]
```diff
@@ -0,0 +1,519 @@
...
+import software.aws.toolkits.telemetry.CodewhispererTelemetry
+import software.aws.toolkits.telemetry.CodewhispererTriggerType
+import software.aws.toolkits.telemetry.Component
+import software.aws.toolkits.telemetry.Result
+import java.time.Duration
+import java.time.Instant
+import java.util.Queue
+
+@Service
+class CodeWhispererTelemetryServiceNew {
```
Comment (Will-ShaoHua) [inline plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererTelemetryServiceNew.kt:56]: doesn't telemetry will share most of the same logic and only have some new param, if only function parameters are different we can just have overloads

File: plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererCodeCoverageTracker.kt [hunk 4]
```diff
@@ -85,25 +89,52 @@ abstract class CodeWhispererCodeCoverageTracker(
...
                         }
                     }
                 }
-            }
-        )
+            )
+        } else {
+            conn.subscribe(
+                CodeWhispererPopupManager.CODEWHISPERER_USER_ACTION_PERFORMED,
+                object : CodeWhispererUserActionListener {
```
Comment (Will-ShaoHua) [inline plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/telemetry/CodeWhispererCodeCoverageTracker.kt:120]: looks like these 2 lines are shared, so we just need one listener with 2 afterAccept.

File: plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/status/CodeWhispererStatusBarWidget.kt [hunk 5]
```diff
@@ -117,7 +119,13 @@ class CodeWhispererStatusBarWidget(project: Project) :
...
         } else if (!isQConnected(project)) {
             AllIcons.RunConfigurations.TestState.Run
-        } else if (CodeWhispererInvocationStatus.getInstance().hasExistingInvocation()) {
+        } else if (
+            if (CodeWhispererFeatureConfigService.getInstance().getNewAutoTriggerUX()) {
+                CodeWhispererInvocationStatusNew.getInstance().hasExistingServiceInvocation()
+            } else {
+                CodeWhispererInvocationStatus.getInstance().hasExistingServiceInvocation()
+            }
+        ) {
```
Comment (Will-ShaoHua) [inline plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/status/CodeWhispererStatusBarWidget.kt:128]: same here if we can do 
```
fun getInstqaonce() = if (newUX) CodeWhispererInvocationStatusNew.getInstance() else service()
```

code here won't need to be changed and just need
```
CodeWhispererInvocationStatus.getInstance().hasExistingServiceInvocation()
```

File: plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/service/CodeWhispererAutoTriggerService.kt [hunk 6]
```diff
@@ -198,7 +178,12 @@ class CodeWhispererAutoTriggerService : CodeWhispererAutoTriggerHandler, Disposa
         var previousOneAccept: Double = 0.0
         var previousOneReject: Double = 0.0
         var previousOneOther: Double = 0.0
-        val previousOneDecision = CodeWhispererTelemetryService.getInstance().previousUserTriggerDecision
+        val previousOneDecision =
+            if (CodeWhispererFeatureConfigService.getInstance().getNewAutoTriggerUX()) {
+                CodeWhispererTelemetryServiceNew.getInstance().previousUserTriggerDecision
+            } else {
+                CodeWhispererTelemetryService.getInstance().previousUserTriggerDecision
```
Comment (Will-ShaoHua) [inline plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/service/CodeWhispererAutoTriggerService.kt:185]: i feel telemetryService can be shared, and have likely maintain 2 previousUserTriggerDecision
```
class CwsprTelemetryService {
      private val previousUserTriggerDecisionOld: Queue<Decision>
      private val previousUserTrigerDecisionNew: Queue<Decision>
     

     val previousUserTriggerDecision: Decision
               get() {
                      if (newUX) {} else {}
                }
}
```

File: plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/service/CodeWhispererAutoTriggerService.kt [hunk 7]
```diff
@@ -84,7 +78,14 @@ class CodeWhispererAutoTriggerService : CodeWhispererAutoTriggerHandler, Disposa
 
     // real auto trigger logic
     fun invoke(editor: Editor, triggerType: CodeWhispererAutomatedTriggerType): Job? {
-        if (!CodeWhispererService.getInstance().canDoInvocation(editor, CodewhispererTriggerType.AutoTrigger)) {
+        if (!(
+                if (CodeWhispererFeatureConfigService.getInstance().getNewAutoTriggerUX()) {
+                    CodeWhispererServiceNew.getInstance().canDoInvocation(editor, CodewhispererTriggerType.AutoTrigger)
+                } else {
+                    CodeWhispererService.getInstance().canDoInvocation(editor, CodewhispererTriggerType.AutoTrigger)
```
Comment (Will-ShaoHua) [inline plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/service/CodeWhispererAutoTriggerService.kt:85]: same 
```
fun getInstance() = if (newUX) newService.getInstance() else service()
```

then we won't need change here

```
CodeWhispererService.getInstance().canDoInvocation(editor, CodewhispererTriggerType.AutoTrigger)
```

File: plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/service/CodeWhispererAutoTriggerHandler.kt [hunk 8]
```diff
@@ -19,7 +19,11 @@ interface CodeWhispererAutoTriggerHandler {
         val triggerTypeInfo = TriggerTypeInfo(CodewhispererTriggerType.AutoTrigger, automatedTriggerType)
 
         LOG.debug { "autotriggering CodeWhisperer with type ${automatedTriggerType.telemetryType}" }
-        CodeWhispererService.getInstance().showRecommendationsInPopup(editor, triggerTypeInfo, latencyContext)
+        if (CodeWhispererFeatureConfigService.getInstance().getNewAutoTriggerUX()) {
+            CodeWhispererServiceNew.getInstance().showRecommendationsInPopup(editor, triggerTypeInfo, latencyContext)
+        } else {
+            CodeWhispererService.getInstance().showRecommendationsInPopup(editor, triggerTypeInfo, latencyContext)
```
Comment (Will-ShaoHua) [inline plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/service/CodeWhispererAutoTriggerHandler.kt:25]: same 
```
fun getInstance() = if (newUX) newService.getInstance() else service()
```

then we won't need change here

```
CodeWhispererService.getInstance().showRecommendationsInPopup(...)
```

File: plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/editor/CodeWhispererEditorListener.kt (hunk 1 above)
Comment (andrewyuq) [reply plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/editor/CodeWhispererEditorListener.kt:37]: CodeWhispererServiceNew.getInstance() is of type `CodeWhispererServiceNew.getInstance()`, so the `getInstance()` will have a type of Any then

**Comment by andrewyuq** [comment]
Most of the comments are around refacting old/new classses into having a common interface and implementations to implement the diffs, there's some work for the refactor, will be pushing the current version as is since it also does the job and revisit the refactor later

**Pull request merged**